src/blockchain/bitcoin_rpc.rs

```diff
@@ -0,0 +1,392 @@
+                let response = self.0.as_mut().unwrap().get_address_info(&address);
+                if let Ok(result) = response {
+                    if let Some(is_watchonly) = result.is_watchonly {
+                        return is_watchonly;
+                    }
+                }
+                false
+            })
+            .any(|x| x);
+        Ok(synced)
+    }
+
+    async fn needs_sync<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        database: &mut D,
+    ) -> Result<(bool, bool), Error> {
+        let synced = match database.get_last_index(ScriptType::External)? {
+            None => false,
+            Some(last_index) => self.index_is_synced(last_index, database).await?,
+        };
+        let needs_rescan = self.index_is_synced(0, database).await?;
+        Ok((!synced, needs_rescan))
+    }
+
+    async fn importmulti<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        rescan_since: Option<u64>,
+        database: &mut D,
+    ) -> Result<(), Error> {
+        let (timestamp, rescan) = match rescan_since {
+            Some(timestamp) => (ImportMultiRescanSince::Timestamp(timestamp), Some(true)),
+            None => (ImportMultiRescanSince::Now, None),
+        };
+        self.0.as_mut().unwrap().import_multi(
+            &database
+                .iter_script_pubkeys(Some(ScriptType::External))?
+                .iter()
+                .map(|script_pubkey| ImportMultiRequest {
+                    script_pubkey: Some(ImportMultiRequestScriptPubkey::Script(&script_pubkey)),
+                    watchonly: Some(true),
+                    timestamp,
+                    ..Default::default()
+                })
+                .collect::<Vec<_>>(),
+            Some(&ImportMultiOptions { rescan }),
+        )?;
+        info!("Successfully imported addresses over bitcoin RPC");
+        Ok(())
+    }
+}
+
+impl Blockchain for BitcoinRpcBlockchain {
+    fn offline() -> Self {
+        BitcoinRpcBlockchain(None)
+    }
+
+    fn is_online(&self) -> bool {
+        self.0.is_some()
+    }
+}
+
+#[async_trait(?Send)]
+impl OnlineBlockchain for BitcoinRpcBlockchain {
+    async fn get_capabilities(&self) -> HashSet<Capability> {
+        vec![].into_iter().collect()
+    }
+
+    async fn setup<D: BatchDatabase + DatabaseUtils, P: Progress>(
+        &mut self,
+        _stop_gap: Option<usize>,
+        database: &mut D,
+        _progress_update: P,
+    ) -> Result<(), Error> {
+        // If node doesn't recognize our most recent address, sync everything
+        // If node doesn't recognize first address, rescan from genesis
+        let (needs_sync, needs_rescan) = self.needs_sync(database).await?;
+        if needs_sync {
+            if needs_rescan {
+                self.importmulti(Some(0), database).await?;
+            } else {
+                self.importmulti(None, database).await?;
+            }
+        }
```

Comment:
Ideally a timestamp could be passed in here to determine how far back Bitcoin Core should rescan. My simple heuristic is "don't rescan if bitcoin core recognizes our first address, do rescan if it doesn't".

Reply:
Done